src/definitions.rs

```diff
@@ -67,10 +67,13 @@ impl<T: Clone + std::fmt::Debug> DefinitionsBuilder<T> {
     pub fn add_definition(&mut self, reference: String, value: T) -> PyResult<ReferenceId> {
         let next_id = self.definitions.len();
         match self.definitions.entry(reference.clone()) {
-            Entry::Occupied(mut entry) => match entry.get_mut().value.replace(value) {
-                Some(_) => py_schema_err!("Duplicate ref: `{}`", reference),
-                None => Ok(entry.get().id),
-            },
+            Entry::Occupied(mut entry) => {
+                // Make the "outermost" schema with a given ref prevail
+                // This means we need to be careful to never generate distinct schemas with the same ref
+                // which would lead to an undetected bug
+                entry.get_mut().value = Some(value);
+                Ok(entry.get().id)
+            }
```

Comment:
I sort of understand the motivation but the fact this silently ignores definitions scares me. It also seems to completely break `definitions_schema`.

I wonder if really the issue is that `ref` keyword to schemas is problematic, and it would be better if instead `definitions_schema` was the only way to create definitions? Basically instead of the `ref` keyword introducing a global definition, the following schemas should be equivalent:

```python
core_schema.date_schema(ref='foobar')

core_schema.definitions_schema(
  core_schema.definition_reference_schema(ref='foobar'),
  {'foobar': core_schema.date_schema()}
)
```

Maybe global scoping of definitions is a consequence of JSONSchema?

Reply:
> It also seems to completely break definitions_schema

How does it break it? 

> I wonder if really the issue is that ref keyword to schemas is problematic, and it would be better if instead definitions_schema was the only way to create definitions?

Really the issue is that in `pydantic` you can have a model like:

```python
class Inner(BaseModel):
    pass
```

That needs to build a non-definitions schema because there is no recursion or multiple references to it. But then that model may be re-used inside another model:

```python
class Outer(BaseModel):
    x: Inner
    y: Inner
```

And now `Inner` needs to become a definition reference schema. And since we're trying to avoid re-generating the schema from scratch and instead start from the basis of `Inner.__pydantic_core_schema__` we need that schema to have a `ref` still, otherwise we wouldn't know if the `x` field is the same as the `y` field.

Reply:
I do think maybe we could make it so that `pydantic-core` only uses a schema in `definitions` if it is within a `definitions` schema.